lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java

```diff
@@ -432,6 +462,12 @@ static TopFieldCollector create(Sort sort, int numHits, FieldDoc after,
       throw new IllegalArgumentException("hitsThresholdChecker should not be null");
     }
 
+    // here we assume that if hitsThreshold was set, we let the corresponding comparator to skip non-competitive docs
+    // It is beneficial for the 1st field only to skip non-competitive docs
+    if (hitsThresholdChecker.getHitsThreshold() != Integer.MAX_VALUE) {
+      sort.fields[0].allowFilterNonCompetitveDocs();
+    }
+
```

Comment:
I think this should be done in `FieldValueHitQueue.create()` below - pass in a boolean to indicate that we want to skip hits if we can, and then do the wrapping/rewriting/whatever it ends up being as a part of building the sort queue

Reply:
@romseygeek  Great suggestion to move wrapping into `FieldValueHitQueue`. I like the APIs more and more now.  Addressed in c84fe5e.
And also looks like we don't even need any changes to `SortField`, we can just wrap comparators.